File: TemplateMethod/Dotnet/ChessGame/Game.cs
namespace ChessGame
{
    public abstract class Game
    {
        protected int currentPlayer;
        protected readonly int _numberOfPlayer;

        public void Run()
        {
            Start();
            while (!HaveWinner)
                TakeTurn();
            Console.WriteLine($"Player {WinningPlayer} wins.");
        }

        protected Game(int numberOfPlayer)
        {
            _numberOfPlayer = numberOfPlayer;
        }

        protected abstract void Start();
        protected abstract void TakeTurn();
        protected abstract bool HaveWinner { get; }
        protected abstract int WinningPlayer { get; }
    }
}

File: TemplateMethod/Python/Chess/game.py
from abc import ABC


class Game(ABC):
    def __init__(self, number_of_players) -> None:
        self.number_of_players = number_of_players
        self.current_player = 0

    def run(self):
        self.start()
        while not self.have_winner():
            self.take_turn()
        print(f"Player {self.winning_player()} wins")

    def start(self) -> None:
        pass

    def have_winner(self) -> bool:
        pass

    def take_turn(self) -> None:
        pass

    def winning_player(self) -> int:
        pass
